content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+
+**Note:** Only one test machine is used in the evaluation of Codestitcher. This immediately raises some red flags. Much of Codestitcher's performance improvements depends upon the memory units such as the instruction cache, unified cache, and TLB. Using multiple machines with different cache configurations and architectures is necessary for a thorough evaluation of Codestitcher as compared to other code layout optimization methods. The machine that was used runs Ubuntu 16.04 on two quad-core 3.60 GHz i7-7700 Kaby-Lake processors. It has a 32 KB L1 instruction cache and 256 KB L2 unified cache.
+
+To test Codestitcher, the authors benchmark five programs with large codebases:
+- MySQL cluster 7.4.12 with the non-transactional read-only test (oltp_simple)
+- Firefox 50.0 with the tp5o benchmark
+- Clang 3.9 with LLVM's multisource benchmark test suite
+- PHP 7.0 (with httpd 2.4) with WP-Test
+- Python 2.7.15 with the apps benchmark group from the Unladen Swallow suite
+
+For each application, the authors compile the binaries and run various benchmarks. The specific compilation configurations used are:
+- PGO: LLVM's default profile-guided optimization
+- PGO.PH: PGO with PH's reordering algorithm applied on top
+- PGO.C3: PGO with call-chain clustering applied on top
+- PH.BB: Baseline binary with PH applied on top
+- CS: Baseline binary with CS applied on top
+
+To do the actual testing, the authors generate profiles over multiple runs of each distinct method, then take the average time improvement for each code layout optimization technique over ten runs.
+
+Here are the results when testing the programs:
+
+![](https://i.imgur.com/tCF1aul.png)
+
+Based on their evaluation, Codestitcher has a visible performance benefit on three of the programs (MySQL, Clang, and Firefox), while PHP and Python show little to no improvement. In fact, most of the code layout optimization methods followed a similar trend. Why is that?
+
+The authors provide an analysis of the MPKI (misses per thousand instructions) on the instruction cache and TLB:
+
+| Application | Instruction cache MPKI | Instruction TLB MPKI |
+|-------------|------------------------|----------------------|
+| MySQL       | 62.44                  | 9.35                 |
+| Clang       | 8.14                   | 1.01                 |
+| Firefox     | 9.16                   | 1.54                 |
+| PHP Server  | 7.63                   | 0.96                 |
+| Python      | 3.40                   | 0.19                 |
+
+As expected, when the load on both the cache and TLB increase, code layout optimization is more effective--- this is evidence that these optimization methods in general are taking advantage of the processor's memory units to improve performance. However, we are not convinced that this is enough evidence to conclude that Codestitcher is superior to PH's method, especially because this is only evaluated on one specific processor.
```

Comment:
```suggestion
As expected, when the load on both the cache and TLB increase, code layout optimization is more effective---this is evidence that these optimization methods in general are taking advantage of the processor's memory units to improve performance. However, we are not convinced that this is enough evidence to conclude that Codestitcher is superior to PH's method, especially because this is only evaluated on one specific processor.
```